base_layer/wallet_ffi/src/lib.rs

```diff
@@ -43,171 +112,361 @@ use tari_comms::peer_manager::NodeIdentity;
+    let key_str;
+    if !key.is_null() {
+        key_str = CStr::from_ptr(key).to_str().unwrap().to_owned();
+    } else {
+        return ptr::null_mut();
+    }
+
+    let public_key = TariPublicKey::from_hex(key_str.as_str());
+    match public_key {
+        Ok(public_key) => Box::into_raw(Box::new(public_key)),
+        Err(_) => ptr::null_mut(),
+    }
+}
 /// -------------------------------------------------------------------------------------------- ///
 
 /// -------------------------------- Private Key ----------------------------------------------- ///
 
+/// Creates a TariPrivateKey from a ByteVector
+///
+/// ## Arguments
+/// `bytes` - The pointer to a ByteVector
+///
+/// ## Returns
+/// `*mut TariPrivateKey` - Returns a pointer to a TariPublicKey. Note that it returns ptr::null_mut()
+/// if bytes is null or if there was an error creating the TariPrivateKey from bytes
 #[no_mangle]
 pub unsafe extern "C" fn private_key_create(bytes: *mut ByteVector) -> *mut TariPrivateKey {
-    let mut v = Vec::new();
+    let v;
     if !bytes.is_null() {
         v = (*bytes).0.clone();
+    } else {
+        return ptr::null_mut();
+    }
+    let pk = TariPrivateKey::from_bytes(&v);
+    match pk {
+        Ok(pk) => Box::into_raw(Box::new(pk)),
+        Err(_) => ptr::null_mut(),
     }
-    let pk = TariPrivateKey::from_bytes(&v).unwrap();
-    Box::into_raw(Box::new(pk))
 }
 
+/// Destroys a TariPrivateKey
+///
+/// ## Arguments
+/// `pk` - The pointer to a TariPrivateKey
+///
+/// ## Returns
+/// `()` - Does not return a value, equivalent to void in C
 #[no_mangle]
 pub unsafe extern "C" fn private_key_destroy(pk: *mut TariPrivateKey) {
     if !pk.is_null() {
         Box::from_raw(pk);
     }
 }
 
+/// Gets a ByteVector from a TariPrivateKey
+///
+/// ## Arguments
+/// `pk` - The pointer to a TariPrivateKey
+///
+/// ## Returns
+/// `*mut ByteVectror` - Returns a pointer to a ByteVector. Note that it returns ptr::null_mut()
+/// if pk is null
 #[no_mangle]
-pub unsafe extern "C" fn private_key_get_byte_vector(pk: *mut TariPrivateKey) -> *mut ByteVector {
+pub unsafe extern "C" fn private_key_get_bytes(pk: *mut TariPrivateKey) -> *mut ByteVector {
     let mut bytes = ByteVector(Vec::new());
     if !pk.is_null() {
         bytes.0 = (*pk).to_vec();
+    } else {
+        return ptr::null_mut();
     }
     Box::into_raw(Box::new(bytes))
 }
 
+/// Generates a TariPrivateKey
+///
+/// ## Arguments
+/// `()` - Does  not take any arguments
+///
+/// ## Returns
+/// `*mut TariPrivateKey` - Returns a pointer to a TariPrivateKey
 #[no_mangle]
 pub unsafe extern "C" fn private_key_generate() -> *mut TariPrivateKey {
     let mut rng = rand::OsRng::new().unwrap();
     let secret_key = TariPrivateKey::random(&mut rng);
     Box::into_raw(Box::new(secret_key))
 }
 
+/// Creates a TariPrivateKey from a char array
+///
+/// ## Arguments
+/// `key` - The pointer to a char array
```

Comment:
add `that holds the key encoded as a hex string`

Reply:
Updated in next commit